### ISSUE TITLE
fp16 training and DeepSpeed support and [QLora for LLAMA-3 70B ]

### ISSUE THREAD

**Issue by SoshyHayami** [pytorch/torchtune#989]
Hi. thank you for bringing this awesome work to us.

If I'm not mistaken, I see that FP16 training is not supported on current recipes for LLama 3, as well as QLora for LLama 3's 70B version.

why does it matter?
because all these three are essential for people who use older, but very common GPU types such as clusters of V100s (32gb) to train their models, especially qlora for the bigger model. unfortunately these GPUs don't have support for newer stuff such as bf16, Flash attention 2, etc. but as you know they're still vram-rich compared to your average consumer-grade cards, so it'd be a waste to ignore them.

As for Deepspeed, I've heard it can offload to CPU ram, so vram may become less of an issue.

Anyway, I don't know if it's part of the plan to add support for these, it'd help tremendously to have them so I appreciate it if you do.

Thanks.

**Comment by kartikayk** [comment]
Thanks so much for opening this issue @SoshyHayami. These are great questions:

> FP16 training is not supported on current recipes for LLama 3

This has been something we've been debating as well. The primary reason we don't support fp16 training is because you really can't train in half-precision with fp16 - in our experience the model doesn't really converge and loss blows up pretty early in training. So you do need to train with mixed precision which has a larger memory footprint since at various points in training you have a copy of the model in fp32 AND fp16. So you roughly 1.5x the peak memory. So far it's been unclear if there is appetite for the intersection of "i have enough memory that I can deal with the 1.5x lift" with "I have GPUs that don't support bf16". From your experience this intersection in fact exists? Would you be able to share more about the setup i.e. do you have 8 V100s and you'd like to train with mixed precision on these?

> QLora for LLama 3's 70B version

We're working through some technical details to enable QLoRA in a distributed setting, namely through interop between FSDP and NF4Tensors in PyTorch. That said, have you already given LoRA a whirl? Or is this blocked by the fp16 support above?

> As for Deepspeed, I've heard it can offload to CPU ram, so vram may become less of an issue

I think we should have support for CPU offload even in FSDP so we dont need deepspeed for this.

cc: @ebsmothers

**Comment by SoshyHayami** [comment]
Thank you very much for the prompt response. 

That's interesting to hear, sorry for the confusion. I was mostly talking about mixed-precision in fp16, not necessarily training the whole thing in half-precision. I must admit that I've not much experience training LLMs this large, but in the audio realm and even the classic encoder-only transformers most mixed-precision setups i've seen so far had the option to use fp16. even the Axolotl which you probably know about, does have the option for it in its Llama 3 configs (but doesn't support multi-gpu training for the full fine tuning). So I was wondering if there's a chance to not load the model in full precision as some people like me don't have the option for bf16, and and my setup is not enough to load the whole thing in fp32.

For now I only have access to 2xV100s, the datacenter I'm using has a couple more V100S so I may be able to get another one as well. whatever I say may not be backed by cold statistics, but based on my own experience, I think access to high-end newer GPUs in the developing countries is not that easy, GPUs such as T4 and V100 strike the right balance of being cheap enough while accessible in large quantity. so that's mostly my two cents.  

I've not tested Lora on Llama 3 70b, I'm sure it won't fit into 64gb of VRAM without quantization. so that's why I hoped to see it being supported here so that maybe alongside offloading to CPU or increasing my setup to 96gb of VRAM, I could finally train it, for now I'm sticking to Llama 3 8B.

Once again, thank you for your support.

**Comment by kartikayk** [comment]
This is great feedback @SoshyHayami!

This makes a lot of sense. Let me add fp16 mixed precision support to our list of TODOs. Or if you'd be open to adding this support to one of the recipes you care about the most, I'd be happy to provide design input and reviews. Let me know if you'd be up for it.

You're right that LoRA 70B will likely not fit into 64GB of VRAM. QLoRA in distributed setting is something we're actively looking into and working with a few teams across PyTorch to enable. Hopefully you'll be able to use this capability very soon!

**Comment by SoshyHayami** [comment]
As much as I'd like to help, I don't think I have the necessary experience with FSDP, though I'll try asking a friend who might be able to help, I'll let you know if I succeeded. 

I'm going to close this thread as I got answer to all my questions. Thank you!
